### PR TITLE
Replace cable_gw_hydro.F90 with its new version

### LINKED ISSUE #230: Replace cable_gw_hydro.F90 with its new version
Replace src/science/gw_hydro/cable_gw_hydro.F90 with its new version as part of the GW work

### PR DESCRIPTION
**Pull request by rkutteh**
# CABLE ground water hydrology work of Mengyuan Mu

This pull request is part of integrating the ground water hydrology work of Mengyuan Mu into CABLE, which will take place in a sequence of pull requests. The actual integration project itself was completed back in September 2023. The current work involves just a gradual merge into the main branch. 

Fixes #230

## Description

1. cable_gw_hydro.F90 is replaced by its new version (drop-in replacement)
2. Dependencies are introduced in **src/offline/cable_parameters.F90**, **src/offline/cable_define_types.F90**, **src/science/soilsnow/cbl_thermal.F90**, and **src/util/cable_common.F90** in order for the code to compile successfully (confirmed)
3. *PLEASE* do not delete any of my in-code comments (tagged rk4417) for now. Once the integration of the ground water hydrology work is complete and its functionality is confirmed to reproduce the results obtained already outside of the repo, I will clean-up all of my comments in a single pull request. Deleting any of my comments now will just make it much more difficult for me to track down and fix any issues that might arise during this process.

----
📚 Documentation preview 📚: https://cable--231.org.readthedocs.build/en/231/

### CODE REVIEW

**Comment by ccarouge** [comment]
@rkutteh I have seen this but I think it is going to take some time. I believe I have to test it against the main version because there are things in there that can change the results without the groundwater on.

**Comment by rkutteh** [comment]
Hi Claire,

no worries, take your time and let me know if you want more info about some of the changes I have made.
We can discuss more on Thursday.

Thanks.
Ramzi
________________________________

File: src/offline/cable_parameters.F90 [hunk 1]
```diff
@@ -1933,7 +2289,7 @@ SUBROUTINE derived_parameters(soil, sum_flux, bal, ssnow, veg, rough)
     END DO
     bal%osnowd0 = ssnow%osnowd
 
-    ! vh_js ! comment out hide% condition
+    !! vh_js !! comment out hide% condition
```
Comment (ccarouge) [inline src/offline/cable_parameters.F90:2292]: No double "!!" please as these are picked up by FORD in the documentation

```suggestion
    ! vh_js ! comment out hide% condition
```

**Review by ccarouge — CHANGES_REQUESTED** [review]
I have finally finished reviewing this. I have a few minor comments, mainly about the documentation and comments as these won't change the results.

There could be a lot more to change but I know there will be some cleanup done once everything is in and we can check the results. And this is "legacy" code we bring in so we can't be too strict.

File: src/science/soilsnow/cbl_thermal.F90 [hunk 2]
```diff
@@ -24,10 +28,12 @@ SUBROUTINE snow_processes_soil_thermal(dels,ssnow,soil,veg,canopy,met,bal,snowml
     TYPE(veg_parameter_type), INTENT(INOUT)  :: veg
     TYPE(met_type), INTENT(INOUT)            :: met ! all met forcing
     TYPE (balances_type), INTENT(INOUT)      :: bal
-    REAL, DIMENSION(:),  INTENT(INOUT)       :: snowmlt
-
+    REAL, DIMENSION(mp)                      :: snowmlt !track snow melt
```
[the remark this answers is not in the document]
Comment (rkutteh) [reply src/science/soilsnow/cbl_thermal.F90:31]: @ccarouge A couple of remarks regarding the requested change:

1. Is the dot in " REAL. " a typo or is it some new syntax I am not aware of ?
2. I realize that the requirement of JULES/UM is a must, but I think it is worth noting that this is not considered good coding practice, as attaching the dimension to the array name is the "old" way of declaring arrays and including the "DIMENSION" attribute in the declaration is the recommended "new" way. I think it is a good idea to double check that this is really what they want so we don't have a repetition of what happened earlier regarding breaking up the modules into smaller ones. 

I am happy to commit the suggestion once I hear back regarding the mysterious "dot"

File: src/science/gw_hydro/cable_gw_hydro.F90 [hunk 3]
```diff
@@ -1129,150 +1213,203 @@ SUBROUTINE calc_equilibrium_water_content(ssnow,soil)
...
 
     END DO
 
   END SUBROUTINE calc_equilibrium_water_content
 
+
   SUBROUTINE calc_srf_wet_fraction(ssnow,soil,met,veg)
+    !* Calculate the wet fraction of the surafce
+    !  following [Decker, 2015](http://doi.wiley.com/10.1002/2015MS000507)
+    !* This subrountine is called in subrountine surf_wetness_fact in cable_canopy.F90
```
Comment (rkutteh) [inline src/science/gw_hydro/cable_gw_hydro.F90:1388]:
```suggestion
  
```

[the remark this answers is not in the document]
Comment (rkutteh) [reply src/science/gw_hydro/cable_gw_hydro.F90]: Done

File: src/offline/cable_parameters.F90 [hunk 4]
```diff
@@ -1765,122 +1779,466 @@ SUBROUTINE derived_parameters(soil, sum_flux, bal, ssnow, veg, rough)
...
+    
+! 2 lines above replaced by below - rk4417 - phase2
+
+    REAL(r_2), DIMENSION(mp,ms) :: psi_tmp, sst_tmp ! MMY add sst_tmp
+    REAL(r_2), DIMENSION(mp,ms) :: soil_depth ! MMY,rhosoil_temp
+
+    REAL(r_2), DIMENSION(:,:), ALLOCATABLE :: ssat_bounded,rho_soil_bulk ! added by rk4417 - phase2
+
+! line below inserted by rk4417 - phase2
+    where(veg%iveg .eq. 17) soil%isoilm = 9   ! MMY@13April it says where iveg = ice, isoilm = permanent ice
```
Comment (rkutteh) [inline src/offline/cable_parameters.F90:1801]:
```suggestion
  
```

File: src/offline/cable_parameters.F90 [hunk 5]
```diff
@@ -1765,122 +1779,466 @@ SUBROUTINE derived_parameters(soil, sum_flux, bal, ssnow, veg, rough)
...
+!    REAL(r_2), DIMENSION(ms) :: soil_depth
+    
+! 2 lines above replaced by below - rk4417 - phase2
+
+    REAL(r_2), DIMENSION(mp,ms) :: psi_tmp, sst_tmp ! MMY add sst_tmp
+    REAL(r_2), DIMENSION(mp,ms) :: soil_depth ! MMY,rhosoil_temp
+
+    REAL(r_2), DIMENSION(:,:), ALLOCATABLE :: ssat_bounded,rho_soil_bulk ! added by rk4417 - phase2
+
+! line below inserted by rk4417 - phase2
```
Comment (rkutteh) [inline src/offline/cable_parameters.F90:1800]:
```suggestion

```

File: src/offline/cable_parameters.F90 (hunk 4 above)
[the remark this answers is not in the document]
Comment (rkutteh) [reply src/offline/cable_parameters.F90]: Done

File: src/science/gw_hydro/cable_gw_hydro.F90 [hunk 6]
```diff
@@ -71,176 +81,187 @@ MODULE cable_gw_hydro_module
...
-    ! if max_ice_frac always set to frozen_limit and tgg_tmp is always CTFRZ
-IMPLICIT NONE
+    SUBROUTINE GWsoilfreeze(dels, soil, ssnow)
+
+   !*## Purpose
+   ! Snow freezes and melts. 
+   ! this is only included because gw_model uses parameters XXX_vec
+   ! these are r_2.  this breaks bitwise compatibility with trunk
+   ! if acceptable this routine does the same thing but with r_2 soil params
+   ! if max_ice_frac always set to frozen_limit and tgg_tmp is always CTFRZ
```
[the remark this answers is not in the document]
Comment (rkutteh) [reply src/science/gw_hydro/cable_gw_hydro.F90:110]: Yes, this subroutine is called only once from inside SUBROUTINE soil_snow_gw (inside cable_gw_hydro.F90). Not sure who put this comment in, but I suggest we leave it until the final clean-up phase of the GW project and worry about it then.

File: src/science/soilsnow/cbl_thermal.F90 (hunk 2 above)
[the remark this answers is not in the document]
Comment (ccarouge) [reply src/science/soilsnow/cbl_thermal.F90:31]: The dot was indeed a typo. But on reflection, I would leave it as is.

**Comment by rkutteh** [comment]
@ccarouge I have resolved the 4 conflicts. I am ready to merge when permitted, so I can then proceed with working on the next pull request.

**Review by ccarouge — APPROVED** [review]
All good to go. Sorry for the wait.

**Comment by rkutteh** [comment]
Thanks Claire. We can cancel our chat today at 12:15 if you wish as I was waiting to merge the pull request to start on what remains, so there is not much to discuss today. However if you still wish to talk please let me know.



Thanks.
Ramzi
________________________________

**Pull request merged**